### ISSUE TITLE
Log output mixed on same line as formatter output

### ISSUE THREAD

**Issue by tyler-ball** [chef/chef#2931]
I have the following test LWRP:

``` ruby
action :default do
  Chef::Log.warn("First log message")
  converge_by("First converge_by") do
    Chef::Log.warn("Second log message")
  end
  converge_by("Second converge_by") do
    Chef::Log.warn("Third log message")
    new_resource.updated_by_last_action(true)
  end
  Chef::Log.warn("Fourth log message")
end
```

When I run this locally, I get the following output:

```
[test-provisioning]$ chef-client -z -o 'test-provisioning::pr275'                                                                                         *[master][ruby-2.1.4@cookbook-test-provisioning]
Starting Chef Client, version 12.1.0.dev.0
[2015-02-17T09:56:14-08:00] WARN: Run List override has been provided.
[2015-02-17T09:56:14-08:00] WARN: Original Run List: []
[2015-02-17T09:56:14-08:00] WARN: Overridden Run List: [recipe[test-provisioning::pr275]]
resolving cookbooks for run list: ["test-provisioning::pr275"]
Synchronizing Cookbooks:
  - test-provisioning
Compiling Cookbooks...
Converging 1 resources
Recipe: test-provisioning::pr275
  * test-provisioning_test_formatter[test] action default[2015-02-17T09:56:14-08:00] WARN: First log message
[2015-02-17T09:56:14-08:00] WARN: Second log message

    - First converge_by[2015-02-17T09:56:14-08:00] WARN: Third log message

    - Second converge_by[2015-02-17T09:56:14-08:00] WARN: Fourth log message

[2015-02-17T09:56:14-08:00] WARN: Skipping final node save because override_runlist was given

Running handlers:
Running handlers complete
Chef Client finished, 1/1 resources updated in 3.291032 seconds
```

The log output is mixed on the same line as the `converge_by` statements.  This is because `converge_by` only calls [resource_update_applied](https://github.com/chef/chef/blob/master/lib/chef/formatters/doc.rb#L259-L274) via [why_run](https://github.com/chef/chef/blob/master/lib/chef/mixin/why_run.rb#L54).  The doc formatter does not apply newlines until [resource_updated](https://github.com/chef/chef/blob/master/lib/chef/formatters/doc.rb#L277-L281) is called.

TLDR: I think we should ensure `Chef::Log` lines appear on their own line when mixed with formatter output.  My proposals:
- Have `resource_update_applied` output entire lines (which end in `\n`)
- Have `Chef::Log` ensure the log line is printing directly after a `\n` character, and if not, add one before printing the log line
- Have `Chef::Log` calls diverted to the formatter when one is enabled.  Then the formatter can make decisions about newlines, indentations, etc.

Thoughts @chef/client-core ?  This has been an issue for me in both Test Kitchen output and Chef-Provisioning output, so I am happy to write the fix.

**Comment by danielsdeleo** [comment]
This isn't really do-able without making some compromises. The problem with option 1("Have resource_update_applied output entire lines") is that you are then prevented from having any additional output on that line. We use that so we can tack on the "up to date" message on the same line when the resource doesn't change. Of course we could buffer that instead, but it makes the output less real-time-y.

If the problem mostly occurs when using the log resource, maybe we could just provide an alternative that uses the formatter or logger intelligently based on the configuration?

**Comment by tyler-ball** [comment]
I actually didn't check the log resource, just direct invocation with `Chef::Log`.  And I don't think switching to only using chef log resources would work for chef-provisioning.

With 1 output stream it doesn't seem possible to have two different processes (formatters and loggers) mix their output intelligently without either communicating to each other or adding buffering.  Maybe the best idea is to just direct logging output to a file all the time, and formatter output to stdout?

**Comment by lamont-granquist** [comment]
There's also this which was slated to get fixed in Chef 12:

https://tickets.opscode.com/browse/CHEF-4874

I'd like to see the logger output largely go away unless someone specifically asks for it.   That does mean that the log resource should really have a converge_by block with some useful output added to it, otherwise that output is invisible when you don't have the logger output.

**Comment by docwhat** [comment]
I know that I currently find the logging output to be really confusing:

* `-l info` doesn't do anything that I can see.
* `-l error` doesn't prevent the normal (formatter, I guess?) output from showing.
* `-l debug -L /tmp/out` still shows all the debugging output to `stdout` and to saves it to `/tmp/out`

Can I suggest the following?

* `-l` by default is `warn`
* `-L` by default is `stderr`
* Continue to send formatter output to `stdout`.
* Add a `--quiet` to turn of the formatter output.
* Only send logging to places specified in `-L`.  Don't send output to multiple places.
* Send output similar to the formatter output to `Chef::Log.info`, but send complete lines (e.g. send the recipe and "(up to date)" all at once).

This should behave a little more like the standards on Unix systems and shouldn't be disruptive to Windows system.

**Comment by lamont-granquist** [comment]
We should definitely do something about this.

FWIW, I kind of think we should drop the real-time-y-ness of the formatter to make things a lot easier and not ever output partial lines.

**Comment by docwhat** [comment]
Getting warnings and errors are important, but they don't have to be real time. As long as they aren't lost. 

Btw: is there documentation for what methods a logger for `log_location` needs? I can't find any. :(

**Comment by josegonzalez** [comment]
+1 for what @docwhat suggested in [this](https://github.com/chef/chef/issues/2931#issuecomment-95663773) comment. It would be lovely to just have a mode where chef says "these are the things that changed" and not need to worry about various formatting options etc. Parsing logs otherwise is a bit annoying.

If anything, we could even separate the `(up to date)` part from the original message.

**Comment by lamont-granquist** [comment]
This should be addressed by #9126

**Comment by lamont-granquist** [comment]
Probably not going to be addressed by #9126 at this point
